applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -0,0 +1,709 @@
+				pvt.sv[i].elevation, pvt.sv[i].azimuth,
+				(pvt.sv[i].flags & NRF_MODEM_GNSS_SV_FLAG_USED_IN_FIX) ? 1 : 0,
+				(pvt.sv[i].flags & NRF_MODEM_GNSS_SV_FLAG_UNHEALTHY) ? 1 : 0);
+		}
+	}
+}
+
+static void on_gnss_evt_fix(void)
+{
+	int err;
+	struct nrf_modem_gnss_pvt_data_frame pvt;
+
+	if (ttft_start != 0) {
+		LOG_INF("TTFF %ds", (int)k_uptime_delta(&ttft_start)/1000);
+		ttft_start = 0;
+	}
+	err = nrf_modem_gnss_read((void *)&pvt, sizeof(pvt), NRF_MODEM_GNSS_DATA_PVT);
+	if (err) {
+		LOG_ERR("Failed to read GNSS PVT data, error %d", err);
+		return;
+	}
+	sprintf(rsp_buf,
+		"\r\n#XGPS: %lf,%lf,%f,%f,%f,%f,\"%04u-%02u-%02u %02u:%02u:%02u\"\r\n",
+		pvt.latitude, pvt.longitude, pvt.altitude,
+		pvt.accuracy, pvt.speed, pvt.heading,
+		pvt.datetime.year, pvt.datetime.month, pvt.datetime.day,
+		pvt.datetime.hour, pvt.datetime.minute, pvt.datetime.seconds);
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+#if defined(CONFIG_SLM_PGPS)
+	struct tm gps_time = {
+		.tm_year = pvt.datetime.year - 1900,
+		.tm_mon  = pvt.datetime.month - 1,
+		.tm_mday = pvt.datetime.day,
+		.tm_hour = pvt.datetime.hour,
+		.tm_min  = pvt.datetime.minute,
+		.tm_sec  = pvt.datetime.seconds,
+	};
+	date_time_set(&gps_time);
+	nrf_cloud_pgps_set_location(pvt.latitude, pvt.longitude);
+#endif /* CONFIG_SLM_PGPS */
+}
+
+#if defined(CONFIG_SLM_AGPS)
+static void agps_req_wk(struct k_work *work)
+{
+	int err;
+	struct gps_agps_request agps_req = {
+	  .sv_mask_ephe = agps_data.sv_mask_ephe,
+	  .sv_mask_alm = agps_data.sv_mask_alm,
+	  .utc = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST) ? 1 : 0,
+	  .klobuchar = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST) ? 1 : 0,
+	  .nequick = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_NEQUICK_REQUEST) ? 1 : 0,
+	  .system_time_tow =
+	      (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST) ? 1 : 0,
+	  .position = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_POSITION_REQUEST) ? 1 : 0,
+	  .integrity = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST) ? 1 : 0
+	};
+
+	err = nrf_cloud_agps_request(agps_req);
+	if (err) {
+		LOG_ERR("Failed to request A-GPS data: %d", err);
+	}
+}
+#endif /* CONFIG_SLM_AGPS */
+
+static void on_gnss_evt_agps_req(void)
+{
+#if defined(CONFIG_SLM_AGPS)
+	int err1 = nrf_modem_gnss_read((void *)&agps_data, sizeof(agps_data),
+					NRF_MODEM_GNSS_DATA_AGPS_REQ);
+	if (err1) {
+		LOG_ERR("Failed to read GNSS PVT data, error %d", err1);
+		return;
+	}
+	k_work_submit(&agps_req);
+#endif /* CONFIG_SLM_AGPS */
+
+#if defined(CONFIG_SLM_PGPS)
+	int err2;
+
+	err2 = nrf_cloud_pgps_request_all();
+	if (err2) {
+		LOG_ERR("Failed to request P-GPS data: %d", err2);
+	}
+	err2 = nrf_cloud_pgps_notify_prediction();
+	if (err2) {
+		LOG_ERR("Failed to request notification of prediction: %d", err2);
+	}
+#endif /* CONFIG_SLM_PGPS */
+}
+
+/* NOTE this event handler runs in interrupt context */
+static void gnss_event_handler(int event)
```

Comment:
I think the implementation should be modified so, that all processing doesn't happen in interrupt context. The log output from `on_gnss_evt_pvt()` and `on_gnss_evt_fix()` probably works because `CONFIG_LOG_MODE_DEFERRED` is enabled, but I would still make those functions work functions and execute those in the a workqueue thread.

`on_gnss_evt_agps_req()` is fine for the A-GPS part (that's already executed in a workqueue), but I don't know how long the P-GPS functions take to execute. If those take long to execute, the code definitely needs to be changed.

Reply:
SLM has its own user workqueue. Will move all event handling to that queue. The READ operation must happen in event handler interrupt context, which should be fine right?

Reply:
Reading data in the event handler is fine, but it's not mandatory. Data can be read also from the workqueue thread, if that's more convenient.

Reply:
All moved to SLM's own work queue.